machine/v1alpha1/types_openstack.go

```diff
@@ -366,3 +372,51 @@ type RootVolume struct {
 	// Deprecated: deviceType will be silently ignored. There is no replacement.
 	DeprecatedDeviceType string `json:"deviceType,omitempty"`
 }
+
+// AdditionalBlockDevice is a block device to attach to the server.
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type != 'local' ? !has(self.availabilityZone) && !has(self.volumeType) : has(self.availabilityZone) || has(self.volumeType)",message="volumeType and availabilityZone must be specified only when type is 'volume'"
+// +union
+type AdditionalBlockDevice struct {
```

Comment:
You've created a bit of an odd embedded union. We advise that each union should only have a single member associated to each discriminant value, and, in general we advise not to create embedded unions, the typical out line for this would be
```
- name: foo
  size: 1
  storage:
    type: Local
```
or
```
- name: foo
  size: 1
  storage:
    type: Volume
    volume:
      type: ...
      availabilityZone: ...
```